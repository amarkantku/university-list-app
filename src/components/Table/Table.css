.table-wrapper {
  margin: 20px 0;
}
table { 
  width: 100%; 
  border-collapse: collapse;
  font-size: 14px;
}
  
th { 
  background: #1abc9c; 
  color: white; 
  font-weight: bold; 
  cursor: cell;
}
td, th { 
  padding: 6px; 
  border: 1px solid #ccc; 
  text-align: left; 
  box-sizing: border-box;
}

tr:nth-of-type(odd) { 
  background: #eee; 
}
tr:hover {background-color:#7fd6c5;}

.no-record {
  margin-top: 2rem;
  font-weight: bold;
  color: rgba(161, 155, 155, 0.712);
}
.clear {
  background: #fff;
  border: none;
}
.hot-link {
  text-decoration: none;
  color: #000;
}
.loading {
  animation: fadeInAnimation ease 3s;
  animation-iteration-count: 1;
  animation-fill-mode: forwards;
}
@keyframes fadeInAnimation {
  0% {
      opacity: 0;
  }
  100% {
      opacity: 1;
  }
}
