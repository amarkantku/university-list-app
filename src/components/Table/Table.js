import React from 'react';
import classNames from 'classnames/bind';
import { Link } from 'react-router-dom';
import Loder from '../common/Loader/Loder';

import styles from './Table.css';
let clx = classNames.bind(styles);

const renderTableHead = (tHead) => {
    return tHead.map(name => <th key={name}>{name}</th>)
}

const renderTableBody = (tBodyData) => {
    return tBodyData.map(university => {
        return (<tr key={university.id}>
            <td>{university.id + 1}</td>
            <td><Link data-testid={`row-name-${university.id}`} className={clx('hot-link')} to={location => ({...location, pathname: `/university/${university.name}`})}>{university.name}</Link></td>
            <td>{university['state-province']}</td>
            <td>{university.country}</td>
            <td>{university.alpha_two_code}</td>
            <td><a data-testid={`row-website-${university.id}`} href={university.web_pages[0]} target='_blank' rel="noopener noreferrer">{university.web_pages[0]}</a></td>
        </tr>)
    })
}

function Table(props) {
    const { tHead, tBodyData, isLoading } = props;
    const isRecordFound = tBodyData.length > 0;
    return (
        <div className={clx('table-wrapper')}>
            <table className={clx('table')}>
                <thead className={clx('thead')}>
                    <tr>{renderTableHead(tHead)}</tr>
                </thead>
                <tbody className={clx('tbody')}>
                    {!isRecordFound && isLoading && (<tr><td className={clx('clear')}><Loder /></td></tr>)}
                    {isRecordFound && !isLoading && renderTableBody(tBodyData)}
                </tbody>
            </table>
            {!isRecordFound && !isLoading && (<div className={clx('no-record')}><p>No Record Found.</p></div>)}
        </div>
    )
}
export default Table;
