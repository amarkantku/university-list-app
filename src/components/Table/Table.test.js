import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { BrowserRouter as Router } from 'react-router-dom';
import Table from './Table';
const defaultProps = {
    tHead: ['name', 'state', 'country', 'website', 'country code'],
    tBodyData: [],
    isLoading: false,
}

describe('<Table />', () => {
    test('component should render', () => {
        const { asFragment } = render(<Table {...defaultProps} />);
        expect(asFragment()).toMatchSnapshot();
    });

    test('should render table with empty record', () => {
        const { getByText, queryAllByText, queryByText } = render(<Table {...defaultProps} />);
        expect(getByText(/name/i)).toBeInTheDocument();
        expect(queryAllByText(/country/i)).toBeTruthy();
        expect(getByText(/website/i)).toBeInTheDocument();
        expect(getByText(/state/i)).toBeInTheDocument();
        expect(queryByText('No Record Found.')).toBeTruthy();
    });

    test('should render with record', () => {
        defaultProps.tBodyData = [
            {
                id: 1,
                name: 'ABC University',
                'state-province': 'Delhi',
                country: 'India',
                alpha_two_code: 'IN',
                web_pages: ['www.abc.test']
            },{
                id: 2,
                name: 'XYZ University',
                'state-province': 'Delhi',
                country: 'India',
                alpha_two_code: 'IN',
                web_pages: ['www.xyx.test']
            }
        ]
        const { getByText, queryAllByText, queryByText } = render(
            <Router>
                <Table {...defaultProps}/>
            </Router>
            );
        expect(getByText(/name/i)).toBeInTheDocument();
        expect(queryAllByText(/country/i)).toBeTruthy();
        expect(getByText(/website/i)).toBeInTheDocument();
        expect(getByText(/state/i)).toBeInTheDocument();
        expect(queryByText('No Record Found.')).toBeFalsy();
        expect(queryByText('XYZ University')).toBeTruthy();
    });

    test('should navigate to other', () => {
        const {getByTestId} = render(
            <Router>
                <Table {...defaultProps}/>
            </Router>
            );
        const divWebsiteEle = getByTestId('row-website-1');
        expect(divWebsiteEle).toBeTruthy();
        fireEvent.click(divWebsiteEle);
        const divNameEle = getByTestId('row-name-1');
        expect(divNameEle).toBeTruthy();
        fireEvent.click(divNameEle);
    });
});