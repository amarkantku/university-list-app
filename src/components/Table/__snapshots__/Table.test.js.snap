// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Table /> component should render 1`] = `
<DocumentFragment>
  <div
    class="table-wrapper"
  >
    <table
      class="table"
    >
      <thead
        class="thead"
      >
        <tr>
          <th>
            name
          </th>
          <th>
            state
          </th>
          <th>
            country
          </th>
          <th>
            website
          </th>
          <th>
            country code
          </th>
        </tr>
      </thead>
      <tbody
        class="tbody"
      />
    </table>
    <div
      class="no-record"
    >
      <p>
        No Record Found.
      </p>
    </div>
  </div>
</DocumentFragment>
`;
